File: Python/Find First and Last Position of Element in Sorted Array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        def binarySearch(nums, left, right, target):
            if left > right:
                return -1
            
            half = int((left + right) / 2)
            if nums[half] < target:
                return binarySearch(nums, half + 1, right, target)
            elif nums[half] == target:
                return half
            else:
                return binarySearch(nums, left, half - 1, target)
        
        index = binarySearch(nums, 0, len(nums) - 1, target)
        if index != -1:
            ans = [index, index]
            for left in range(index - 1, -1, -1):
                if nums[left] == target:
                    ans[0] = left
                else:
                    break
            for right in range(index + 1, len(nums)):
                if nums[right] == target:
                    ans[1] = right
                else:
                    break
            return ans
        
        return [-1, -1]

File: C++/Find First and Last Position of Element in Sorted Array.cpp
class Solution {
private:
    // binary search for sorted array
    int binarySearch(vector<int>& nums, int target, int left, int right){
        if(left > right)
            return -1;
        
        int mid = (int) ((left + right) / 2);
        
        if(nums[mid] == target)
            return mid;
        else if(nums[mid] > target)
            return binarySearch(nums, target, left, mid - 1);
        else
            return binarySearch(nums, target, mid + 1, right);
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        
        int pos = binarySearch(nums, target, 0, nums.size() - 1);
        
        if(pos == -1){
            ans.push_back(-1);
            ans.push_back(-1);
        }
        else{
            // find the boundary of target value
            int left = 0, right = 0;
            while(pos - left >= 0 and nums[pos - left] == target)
                left++;
            
            while(pos + right < nums.size() and nums[pos + right] == target)
                right++;
            
            ans.push_back(pos - left + 1);
            ans.push_back(pos + right - 1);
        }
        
        
        return ans;
    }
};